builder/gridscale/builder.go

```diff
@@ -0,0 +1,147 @@
+package gridscale
+
+import (
+	"context"
+	"fmt"
+	"github.com/hashicorp/hcl/v2/hcldec"
+	"log"
+
+	"github.com/gridscale/gsclient-go/v3"
+
+	"github.com/hashicorp/packer/common"
+	"github.com/hashicorp/packer/helper/communicator"
+	"github.com/hashicorp/packer/helper/multistep"
+	"github.com/hashicorp/packer/packer"
+)
+
+// The unique id for the builder
+const BuilderId = "packer.gridscale"
+
+type Builder struct {
+	config Config
+	runner multistep.Runner
+}
+
+func (b *Builder) ConfigSpec() hcldec.ObjectSpec { return b.config.FlatMapstructure().HCL2Spec() }
+
+func (b *Builder) Prepare(raws ...interface{}) ([]string, []string, error) {
+	c, warnings, errs := NewConfig(raws...)
+	if errs != nil {
+		return nil, warnings, errs
+	}
+	b.config = *c
+
+	return nil, nil, nil
+}
+
+func (b *Builder) Run(ctx context.Context, ui packer.Ui, hook packer.Hook) (packer.Artifact, error) {
+	client := gsclient.NewClient(gsclient.DefaultConfiguration(b.config.APIKey, b.config.APIToken))
+	// Build the steps
+	steps := []multistep.Step{
+		&stepCreateServer{
+			client: client,
+			config: &b.config,
+			ui:     ui,
+		},
+		&stepCreateSSHKey{
+			Debug:        b.config.PackerDebug,
+			DebugKeyPath: fmt.Sprintf("gs_%s.pem", b.config.PackerBuildName),
+			client:       client,
+			config:       &b.config,
+			ui:           ui,
+		},
+		&stepCreateBootStorage{
+			client: client,
+			config: &b.config,
+			ui:     ui,
+		},
+		&stepLinkServerBootStorage{
+			client: client,
+			config: &b.config,
+			ui:     ui,
+		},
+		&stepCreateSecondaryStorage{
```

Comment:
`SecondaryStorage` should be optional!

Reply:
Yes, it is optional in the step implementation.

Reply:
Check the step go file, you will see it is optional.

Reply:
Ok.